src/main/java/org/openlmis/buq/service/buq/BottomUpQuantificationService.java

```diff
@@ -693,30 +693,39 @@ private List<ProductGroupsCostData> createProductsCostData(boolean isDistrictLev
       }
     } else {
       for (UUID locationId : subZones) {
-        ProductGroupsCostData productsCosts = new ProductGroupsCostData();
-        productsCosts.setDataSourceId(locationId);
         List<MinimalFacilityDto> facilityDtos = facilityReferenceDataService.search(
                 null, null, locationId, true).stream()
             .filter(this::checkFacilityTypeAndPermission)
             .collect(Collectors.toList());
+        Set<String> facilityTypes = facilityDtos.stream()
+            .map(dto -> dto.getType().getName())
+            .collect(Collectors.toSet());
+
         Set<UUID> facilityIds = facilityDtos.stream()
             .map(BaseDto::getId).collect(Collectors.toSet());
 
-        List<BottomUpQuantification> bottomUpQuantificationsForCalculations =
-            bottomUpQuantificationList.stream()
-                .filter(buq -> facilityIds.contains(buq.getFacilityId()))
-                .collect(Collectors.toList());
-
-        Map<String, String> calculatedGroups =
-            calculateProductGroupsCost(bottomUpQuantificationsForCalculations);
-        productsCosts.setCalculatedGroupsCosts(calculatedGroups);
-
-        List<UUID> bottomUpQuantificationsForCostCalculationIds =
-            bottomUpQuantificationList.stream()
-                .map(BottomUpQuantification::getId)
-                .collect(Collectors.toList());
-        productsCosts.setBottomUpQuantificationIds(bottomUpQuantificationsForCostCalculationIds);
-        productsCostsList.add(productsCosts);
+        for (String facilityType : facilityTypes) {
+          List<BottomUpQuantification> bottomUpQuantificationsForCalculations =
+              bottomUpQuantificationList.stream()
+                  .filter(buq -> facilityIds.contains(buq.getFacilityId()))
+                  .filter(buq -> getFacility(facilityDtos, buq.getFacilityId()).getType().getName()
+                      .equals(facilityType))
```

Comment:
I think there is no need to do double filter, because you should be able to just add `&&` in the first `.filter` and include this statement there. By doing so, you will only loop the array once, not twice.

Reply:
Changed it to
.filter(buq -> facilityIds.contains(buq.getFacilityId())
                      && getFacility(facilityDtos,